### PR TITLE
app-shells/scsh: multiple fixes

### PR DESCRIPTION
**Pull request by xgqt**
- apply patch from adam@michalik.es presented in bug #413791
- patch ar call in Makefile
- bump EAPI to 8
- style changes

Closes: https://bugs.gentoo.org/722442
Closes: https://bugs.gentoo.org/413791
Package-Manager: Portage-3.0.20, Repoman-3.0.3
Signed-off-by: Maciej Barć <xgqt@riseup.net>

### CODE REVIEW

**Comment by xgqt** [comment]
@gktrk NetBSD's pkgsrc lists scsh as "Not 64-bit clean.", so https://bugs.gentoo.org/589122#c5 probably can't be solved.

http://cvsweb.netbsd.org/bsdweb.cgi/~checkout~/pkgsrc/shells/scsh/Makefile?rev=1.46&content-type=text/plain

**Comment by xgqt** [comment]
Scsh's unofficial version 0.7 compiles but doesn't run

image

**Comment by gktrk** [comment]
> Scsh's unofficial version 0.7 compiles but doesn't run
> 
> image

I loosely followed this: https://web.archive.org/web/20140902224032/http://thread.gmane.org/gmane.lisp.scheme.scsh/7821 Used dev-scheme/scheme48-1.9.2-r1. Looks like the unofficial 0.7 works:
Screenshot_2021-11-09_00-03-37

They seem to do a crappy job with handling file paths. Run it with `strace -f -s 4096 /usr/bin/scsh` and look for a failing execve like this:
```shell
execve("/usr/local/lib/scheme48-1.9.2/scheme48vm", ["/usr/local/bin/scsh", "-I", "/usr/local/lib/scsh-0.7/scsh.image", "/usr/local/bin/scsh"], 0x7ffed6ce1328 /* 9 vars */) = -1 ENOENT (No such file or directory)
```

Your paths will look different but the idea is that it's trying to execve a non-existing binary. If you fix that, it should drop to the shell like it did with mine.

The bigger question though is what you'd do about this if you were a Gentoo developer. How do you propose that we should proceed with handling this bug? It must be resolved one way or another.

**Comment by xgqt** [comment]
AD 0.6.X
@gktrk 

> The bigger question though is what you'd do about this if you were a Gentoo developer. How do you propose that we should proceed with handling this bug? It must be resolved one way or another.

If we absolutely have to have it in the tree I would mask it on multilib and add correct abi32 deps.
If we don't then I would last-rite it.

**Comment by gktrk** [comment]
> AD 0.6.X @gktrk
> 
> > The bigger question though is what you'd do about this if you were a Gentoo developer. How do you propose that we should proceed with handling this bug? It must be resolved one way or another.
> 
> If we absolutely have to have it in the tree I would mask it on multilib and add correct abi32 deps. If we don't then I would last-rite it.

Let's go with the option of keeping it. Add the necessary masks and deps to this PR. That should about wrap this up, the bugs this PR fixes are sufficient.

**Comment by gktrk** [comment]
> AD 0.7.X
> 
> I wasnt able to patch scsh build system to install the IMAGE with correct paths (scsh.image still had the tempdir appended when installed with `insinto`). So, I had split it into two pkgs one installs the files for image (`scsh-lib`) and then the other (`scsh`) can embed the corrwect system paths into the image. I have no idea why 0.7 behaves like that, 0.6 worked correctly.

At a quick glance they look OK. I'll runtest it today and see if we can justify introducing another package. It might as well turn out to be the correct way.

**Review by gktrk — CHANGES_REQUESTED** [review]
I've spotted a few things. I'm also investigating the need for scsh-lib. Will follow up with that later.

File: app-shells/scsh/scsh-0.6.7-r3.ebuild
Comment (xgqt) [reply app-shells/scsh/scsh-0.6.7-r3.ebuild:21]: slot is already added. would portage error on this?

File: app-shells/scsh/scsh-0.6.7-r3.ebuild
Comment (xgqt) [reply app-shells/scsh/scsh-0.6.7-r3.ebuild:16]: we're fixing the old one here (what was stable and had build failures)

Comment (gktrk) [reply app-shells/scsh/scsh-0.6.7-r3.ebuild:16]: Right but you're changing `DEPEND`, and consequently `RDEPEND` in-place. Remember the Portage static vs. dynamic dependency model?

File: profiles/arch/amd64/no-multilib/package.mask
Comment (gktrk) [inline profiles/arch/amd64/no-multilib/package.mask:26]: Can you tell me why `<app-shells/scsh-0.7.0` didn't work? What was wrong with it?

Comment (xgqt) [reply profiles/arch/amd64/no-multilib/package.mask:26]: > Can you tell me why `<app-shells/scsh-0.7.0` didn't work? What was wrong with it?

becasue _pre was lesser than full version

File: app-shells/scsh/scsh-0.6.7-r3.ebuild
Comment (gktrk) [reply app-shells/scsh/scsh-0.6.7-r3.ebuild:21]: Portage recorded the deps on my machine as `!dev-scheme/scheme48 virtual/libcrypt:0/1= virtual/libcrypt[abi_x86_32]`. Since `:0/1=` is stricter, it should eliminate others pulled in by the non-slot variant. One could assume that portage merges the USE flag in the non-slot variant with the slotted variant as well. If it were me however, I would use the slot operator on both deps to be on the safe side.

Comment (xgqt) [reply app-shells/scsh/scsh-0.6.7-r3.ebuild:21]: sure

File: profiles/arch/amd64/no-multilib/package.mask
Comment (gktrk) [reply profiles/arch/amd64/no-multilib/package.mask:26]: > > Can you tell me why `<app-shells/scsh-0.7.0` didn't work? What was wrong with it?
> 
> becasue _pre was lesser than full version

That was my initial reaction too. Then I copied the same exact ebuild as `app-shells/scsh-0.7_p20160820` and got the following error:

```shell
!!! All ebuilds that could satisfy "scsh" have been masked.
!!! One of the following masked packages is required to complete your request:
- app-shells/scsh-0.7_p20160820::gentoo (masked by: package.mask)
/var/db/repos/gentoo/profiles/arch/amd64/no-multilib/package.mask:
# AMD64 Team <amd64@gentoo.org>
# Mask packages that rely on amd64 multilib

- app-shells/scsh-0.7_pre20160820::gentoo (masked by: package.mask)
- app-shells/scsh-0.6.7-r3::gentoo (masked by: package.mask)
```

Any guesses as to why?

Comment (xgqt) [reply profiles/arch/amd64/no-multilib/package.mask:26]: no idea

Comment (gktrk) [reply profiles/arch/amd64/no-multilib/package.mask:26]: So the ebuild was `0.7_pre20160820` and the mask was `0.7.0`. Take a look at https://dev.gentoo.org/~ulm/pms/head/pms.html#x1-26008r2 and see if you can figure it out.

Comment (xgqt) [reply profiles/arch/amd64/no-multilib/package.mask:26]: the more "numeric components" the greater a package version would be considered,
so even if we mask `0.7.0`, `0.7` would be lower

Comment (gktrk) [reply profiles/arch/amd64/no-multilib/package.mask:26]: Right on point, awesome 👍

**Comment by gktrk** [comment]
Given their braindead rule in `Makefile.in`:

```Makefile
install-scsh-image: install-scsh
        $(srcdir)/build/build-image.sh $(srcdir) \
                "$(LIB)/" '$(DESTDIR)$(LIB)/scsh.image' '$(SCHEME48) -h 0' '$(LOADS)'
```

where library path stored in `$(LIB)` is assumed to be relative to `/` but the image is installed under `$(DESTDIR)`, it seems that upstream assumes libs are merged on the system before the image is built.

In our case though, we want to build the image along with the libs together. So I tried sed'ing that rule so that it looks like:

```Makefile
install-scsh-image: install-scsh
        $(srcdir)/build/build-image.sh $(srcdir) \
                "$(DESTDIR)""$(LIB)/" '$(DESTDIR)$(LIB)/scsh.image' '$(SCHEME48) -h 0' '$(LOADS)'
```

reading the libraries from `$(DESTDIR)` as well. It builds just fine but due to their very fragile path handling, scsh fails after merge with:

```shell
$ scsh
VM exception `external-error' with no handler in place
opcode is: call-external-value-2
stack template id's:  <- 293 <-  <-  <- 1258 <- 1258 <- 6650 <- 6625 <- 3309 <- 
```

I have the feeling that `build-image.sh` hardcodes the path to the libs, so after a merge it's still looking for it under `${PORTAGE_TMPDIR}`. Given that, I think the 2-package approach is justified here.

File: app-shells/scsh/scsh-0.6.7-r3.ebuild
Comment (gktrk) [inline app-shells/scsh/scsh-0.6.7-r3.ebuild:13]: Bump `HOMEPAGE` to https

File: app-shells/scsh/scsh-0.6.7-r3.ebuild
Comment (gktrk) [inline app-shells/scsh/scsh-0.6.7-r3.ebuild:20]: Speaking of this libcrypt dependency, could you remind me again why it's needed? I grepped the source code for `crypt` but found nothing. I ran scanelf on every executable but didn't see a dependency for libcrypt.

**Review by gktrk — CHANGES_REQUESTED** [review]

File: app-shells/scsh/scsh-0.6.7-r3.ebuild
Comment (xgqt) [reply app-shells/scsh/scsh-0.6.7-r3.ebuild:20]: It links to `crypt`
```
/srfi/srfi-27.o c/unix/misc.o c/unix/io.o c/unix/fd-io.o c/unix/event.o         \
        /tmp/s48_external_$$.c                                  \
        c/unix/socket.o c/unix/dynamo.o                         \
         -lcrypt -ldl -lm -lc -lutil && 
```
Also, if this is removed portage complains:
```
* QA Notice: Binary files linked to libcrypt.so found
 * But a dependency on virtual/libcrypt is not declared!
 *
 *   /var/tmp/portage/app-shells/scsh-0.6.7-r3/image/usr/lib/scsh/scshvm
 *
 * Please add virtual/libcrypt:= dependency
 ```

Comment (thesamesam) [reply app-shells/scsh/scsh-0.6.7-r3.ebuild:20]: In 0.6.7 (already in tree), I see e.g.:
```
scsh/syscalls1.c:1109:/* crypt()
scsh/syscalls1.c:1112:s48_value scm_crypt(s48_value key, s48_value salt)
scsh/syscalls1.c:1114:  char * ret = crypt (s48_extract_string (key),
scsh/syscalls1.c:1209:  S48_EXPORT_FUNCTION(scm_crypt);
```

so it seems correct.

File: app-shells/scsh-lib/scsh-lib-0.7_pre20160820.ebuild
Comment (xgqt) [reply app-shells/scsh-lib/scsh-lib-0.7_pre20160820.ebuild:34]: you are right, it is compiled with CC, need to patch this then

Comment (thesamesam) [reply app-shells/scsh-lib/scsh-lib-0.7_pre20160820.ebuild]: The general lesson here being: please don't rush to put things into QA_FLAGS_IGNORED. It's not just a QA issue but can break things on various arches and indeed cross compilation. But I totally get being misled by Scheme stuff too, it was plausible that it was some sort of its own compiler.

File: app-shells/scsh/scsh-0.7_pre20160820.ebuild
Comment (thesamesam) [reply app-shells/scsh/scsh-0.7_pre20160820.ebuild:22]: If we can't fix that one, we should try skip it so that we run _something_ to know if it's working during arch testing and such.

File: app-shells/scsh/scsh-0.7_pre20160820.ebuild
Comment (thesamesam) [inline app-shells/scsh/scsh-0.7_pre20160820.ebuild:61]: I'd probably derive this path from the ebuild version using the version manipulation commands. Ditto the next line.

File: profiles/arch/amd64/no-multilib/package.mask
Comment (thesamesam) [reply profiles/arch/amd64/no-multilib/package.mask:26]: Why this does this actually need multilib again?

File: app-shells/scsh/scsh-0.7_pre20160820.ebuild
Comment (xgqt) [reply app-shells/scsh/scsh-0.7_pre20160820.ebuild:58]: because we need `install` from Makefile to correct the image's load paths. It has to install to some path from which we then extract `scsh.image`.

**Comment by xgqt** [comment]
> Why this does this actually need multilib again?

because 0.6.7 only compiles on 32-bits

File: app-shells/scsh/scsh-0.6.7-r3.ebuild
Comment (xgqt) [reply app-shells/scsh/scsh-0.6.7-r3.ebuild:16]: https://github.com/gentoo/gentoo/pull/22703#discussion_r749617041

File: app-shells/scsh/scsh-0.6.7-r3.ebuild
Comment (xgqt) [reply app-shells/scsh/scsh-0.6.7-r3.ebuild:6]: Yes, we do need it
```
/var/tmp/portage/app-shells/scsh-0.6.7-r3/temp/environment: line 579: multilib_toolchain_setup: command not found
```

File: app-shells/scsh-lib/files/Makefile.in-LDFLAGS.patch
Comment (xgqt) [reply app-shells/scsh-lib/files/Makefile.in-LDFLAGS.patch]: fixed

File: app-shells/scsh/scsh-0.7_pre20160820.ebuild
Comment (xgqt) [reply app-shells/scsh/scsh-0.7_pre20160820.ebuild:61]: meh, more complicated + will this pkg be kept in the tree after this?

File: app-shells/scsh/scsh-0.6.7-r3.ebuild
Comment (gktrk) [reply app-shells/scsh/scsh-0.6.7-r3.ebuild:20]: You're right, 0.6.7 has it. I only checked the source for 0.7 and that one doesn't have any references to crypt.

File: app-shells/scsh/scsh-0.7_pre20160820.ebuild
Comment (gktrk) [reply app-shells/scsh/scsh-0.7_pre20160820.ebuild:59]: That's just upstream sillyness. If you look at `Makefile.in`:

```Makefile
install-scsh: scsh
...
        for file in $(patsubst c/%,%,$(COMPILED_LIBS)) ; do \
        $(INSTALL) c/$$file $(DESTDIR)$(LIB)/$$file ; \
        done
...
```

it's installing the lib files to `$(DESTDIR)$(LIB)`. Now look at the build rule for the image:
```Makefile
install-scsh-image: install-scsh
        $(srcdir)/build/build-image.sh $(srcdir) \
                "$(LIB)/" '$(DESTDIR)$(LIB)/scsh.image' '$(SCHEME48) -h 0' '$(LOADS)'
```

it's not pulling the libs from `$(DESTDIR)`, it's relative to `/` now. So that necessitates the scsh-lib package, which installs these libs first. Take a look again at the prerequisite for the rule, it depends on `install-scsh`. Which is dumb, because it's not consuming the output of `install-scsh`, unless your DESTDIR is `/`. So you can't call `install-scsh-image` without introducing a file collision with `scsh-lib`. Thus, you let it install to a temporary location, and pick out `scsh.image` out of that.

Alternatively, you can sed out the prerequisite with:
```sh
sed -e '/install-scsh-image:/ s|:.*|:|' -i Makefile.in
```
in which case you don't need to use a temporary dir, you can just use `DESTDIR="${D}"`

File: app-shells/scsh/scsh-0.7_pre20160820.ebuild
Comment (gktrk) [reply app-shells/scsh/scsh-0.7_pre20160820.ebuild]: > ~meh, more complicated + will this pkg be kept in the tree after this?~
> will do that anyway

Right, it's important to keep in mind that this recruitment task is to test whether you are willing to do your due diligence when handling packages or not. Better to over-deliver than under-deliver. It sure is more work than your typical maintenance task but it shows me that if you demonstrate putting the necessary effort here, you'll do well with simpler tasks in the future.

Also, I don't see why it must be treecleaned if we fix it and it's working.

File: app-shells/scsh-lib/scsh-lib-0.7_pre20160820.ebuild
Comment (gktrk) [reply app-shells/scsh-lib/scsh-lib-0.7_pre20160820.ebuild]: What do we need to patch this with? Is it causing any QA violations?

File: app-shells/scsh/scsh-0.7_pre20160820.ebuild
Comment (gktrk) [reply app-shells/scsh/scsh-0.7_pre20160820.ebuild]: The test script returns 0 even in this case, because upstream ...

File: app-shells/scsh/scsh-0.6.7-r3.ebuild
Comment (gktrk) [reply app-shells/scsh/scsh-0.6.7-r3.ebuild:16]: Sam's point is that you can add -r3 without dropping the stable -r2

Comment (xgqt) [reply app-shells/scsh/scsh-0.6.7-r3.ebuild:16]: > Sam's point is that you can add -r3 without dropping the stable -r2

but there is no way that -r2 could have been installed on any system because of the AR bug

File: app-shells/scsh-lib/scsh-lib-0.7_pre20160820.ebuild
Comment (xgqt) [reply app-shells/scsh-lib/scsh-lib-0.7_pre20160820.ebuild]: > What do we need to patch this with? Is it causing any QA violations?

yes, not respecting `LDFLAGS`, I added that to Makefile and as @thesamesam suggested I added also `CPPFLAGS`
https://github.com/gentoo/gentoo/pull/22703/commits/036424b012a07d3bb4a67639bed59c5c1cd26526#diff-7c6fec4e5fd9a2d2f098adce1f45a9dc73ecd25addd13eee5c29574ea1a6e41fR8

File: app-shells/scsh/scsh-0.6.7-r3.ebuild
Comment (gktrk) [reply app-shells/scsh/scsh-0.6.7-r3.ebuild:16]: > > Sam's point is that you can add -r3 without dropping the stable -r2
> 
> but there is no way that -r2 could have been installed on any system because of the AR bug

I see your point but someone did stabilize this ebuild. And before that another developer added this ebuild. Chances of two developers okaying this despite a clear build failure is very low I think. Unless, the two were the same developer, which appears to be the case here unfortunately ... You are likely to be right in this case, though it's simply safer to leave the stable version alone and revbump to unstable.

**Review by gktrk — APPROVED** [review]
As far as recruitment is concerned, this is good for me. My own testing The commits individually are not clean, and they are missing `Closes:` tags but I'll leave it to @gentoo/lisp to decide how they want it addressed.

@xgqt We'll do one short session on IRC when you're available to discuss how this task went and finalize the recruitment.

**Comment by xgqt** [comment]
0.6.7 done via
https://gitweb.gentoo.org/repo/gentoo.git/commit/?id=93fbed6e982f91eb32076fbf42c017cfcdd6c7ac

**Comment by xgqt** [comment]
0.7 done via
https://gitweb.gentoo.org/repo/gentoo.git/commit/?id=ca8c43d3d9c60c241f8d91205901ec3de7d100ca

**Pull request closed without merging**
